File: Sales_P3/src/pandas/0_merge_copa_sales.py
import pandas as pd
from pathlib import Path
from janitor import clean_names


# Path
path = Path(__file__).parent.parent.parent


# Functions
def read_multiple_files(list_of_files):
    dataframes = [
        pd.read_csv(
            file,
            delimiter="\t",
            skiprows=10,
            encoding="UTF-16LE",
            skipinitialspace=True,
            thousands=",",
            engine="python",
            dtype={
                "Period": str,
                "CoCd": str,
                "Doc. no.": str,
                "Ref.doc.no": str,
                "AC DocumentNo": str,
                "Delivery": str,
                # "Item": int,
                "Plnt": str,
                "Tr.Prt": str,
                "ConsUnit": str,
                "FIRE Plant": str,
                "FIREOutlet": str,
            },
        ).rename(
            # Rename columns that are changing between data source
            columns={"Stock val.": "Stock Value"}
        )
        for file in list_of_files
    ]

    # Merge the list of DataFrames into a single DataFrame
    df = pd.concat(dataframes)

    return df


def remove_first_two_columns(df):
    return df.iloc[:, 2:]


def remove_sub_total_rows(df):
    return df.loc[df["RecordType"].notna()]


def clean_column_names(df):
    df = df.clean_names()
    df.columns = df.columns.str.strip("_")
    return df


def main():
    # Variable
    year = 2024

    # Filenames
    data_path = path / "data" / "Actual"
    output_file = path / "db" / f"COPA_Sales_{year}.parquet"

    # Input data: List of multiple text files
    txt_files = [
        file for file in data_path.iterdir() if file.is_file() and file.suffix == ".TXT"
    ]

    # Process data
    df = read_multiple_files(txt_files)
    df = (
        df.pipe(remove_first_two_columns)
        .pipe(remove_sub_total_rows)
        .pipe(clean_column_names)
    )

    # Write to Parquet file
    df.to_parquet(output_file, index=False)
    print("A parquet file is created.")


if __name__ == "__main__":
    main()

File: Sales_P3/src/dplyr/0_merge_copa_sales.R
library(dplyr)
library(readr)
library(arrow)
library(purrr)
library(fs)
library(janitor)
library(here)
library(glue)


# Path
path <- here("Sales_P3")


# Functions
read_txt_file <- function(file_path) {
  # Read tsv file
  df <- read_tsv(
    file_path,
    skip = 10,
    locale = locale(encoding = "UTF-16LE"),
    col_types = cols(
      "Period" = col_character(),
      "CoCd" = col_character(),
      "Doc. no." = col_character(),
      "Ref.doc.no" = col_character(),
      "AC DocumentNo" = col_character(),
      "Delivery" = col_character(),
      # "Item" = col_integer(),
      "Plnt" = col_character(),
      "Tr.Prt" = col_character(),
      "ConsUnit" = col_character(),
      "FIRE Plant" = col_character(),
      "FIREOutlet" = col_character(),
    )
  )
  # Correct the occasional column name in the data source
  if ("Stock val." %in% names(df)) {
    df <- df |>
      rename(c("Stock Value" = "Stock val."))
  }
  return(df)
}


read_multiple_files <- function(list_of_files) {
  df <- list_of_files |>
    map_dfr(read_txt_file)
  return(df)
}


remove_first_two_columns <- function(df) {
  df <- df |>
    select(!c("...1", "...2"))
  return(df)
}

remove_unnecessary_columns <- function(df) {
  # Remove unnecessary columns which are added since 2024-04 data
  df <- df |>
    select(!c("AAG", "MatNr Val. Class"))
}


remove_sub_total_rows <- function(df) {
  df <- df |>
    filter(!is.na(df$RecordType))
  return(df)
}


clean_column_names <- function(df) {
  return(df)
}


main <- function() {
  # Variable
  year <- 2024

  # Path
  data_path <- here(path, "data", "actual")

  # Filenames
  input_files <- dir_ls(data_path, regexp = "\\.TXT")
  output_file <- here(path, "db", glue("COPA_Sales_{year}.parquet"))

  # Read data
  df <- read_multiple_files((input_files))

  # Process data
  df <- df |>
    remove_first_two_columns() |>
    remove_unnecessary_columns() |>  # which are added since 2024-04 data
    remove_sub_total_rows() |>
    clean_names()

  # Write data
  write_parquet(df, output_file)
  print("A parquet file is created")
}

main()
